core/drivers/tzc380.c

```diff
@@ -170,7 +170,7 @@ static uint32_t addr_high(vaddr_t addr __maybe_unused)
  * `tzc_configure_region` is used to program regions into the TrustZone
  * controller.
  */
-void tzc_configure_region(uint8_t region, vaddr_t region_base, uint32_t attr)
```

Comment:
What's conflicting with `uint32_t`?

Reply:
Yes, I see compilation error with it.

Reply:
To me it looks like `attr` is supposed to be 32 bits wide. Perhaps you're only fixing the symptom?

Reply:
Actually, looked at the vendor-specific code as well, https://source.codeaurora.org/external/imx/imx-optee-os/tree/core/drivers/tzc380.c?h=imx_4.9.88_2.0.0_ga#n149 that also suggest attr should be of type size_t

Reply:
Or there's an accident with a signed type somewhere...
This has to be understood better.

Reply:
if you look at, https://github.com/OP-TEE/optee_os/blob/master/core/include/drivers/tzc380.h#L200 , type here is size_t which is conflicting with uint32_t ?

Reply:
That seems to be the problem then, that the prototype of the function is wrong.

Reply:
`size_t` is definitely not the right type, as it has nothing to do with the *virtual* address size. The actual register written to is a 32-bit one, so eventually it could to become a uint32_t at some point. But the TZC relates to physical memory, and supports 64-bit buses, so I guess a paddr_t or uint64_t would be the type to go for.
But actually the current interface is somewhat broken, as the prototype does not match the implementation, both in terms of naming and semantics(!):
`void tzc_configure_region(uint8_t region, vaddr_t region_base, size_t size);`
vs.
`void tzc_configure_region(uint8_t region, vaddr_t region_base, uint32_t attr) {...`

So I would guess it should become:
`void tzc_configure_region(uint8_t region, paddr_t region_base, paddr_t size);`
(I think there is no psize_t type).
And then we would need code to convert the size, presumably given in bytes, to the power-of-two encoding used in the register. This would then be handed down to tzc_write_region_attributes in an uint32_t.
But without an actual user it's hard to tell what's the right thing here.

Reply:
@Andre-ARM , 
i will disagree with your conclusion. (although the start is correct)
attr variable (the third parameter in this function) represents the value that will be written to the Region Attributes register  that is as you have written a 32 bits register. This is true regardless of the ARM architecture (v7 or v8) or the width of the bus it connects too. 
it is also true that there is a discrepency between the prototype and the function implemented.
but it is the function that is correct and the prototype a mistake (probably from a previous implementation)
The conclusion should be that the prototype should be fixed to match the function
and so:
`void tzc_configure_region(uint8_t region, vaddr_t region_base, uint32_t attr);`

Reply:
Fair enough, if handing over an attribute was really the intention. In this case the type should be uint32_t indeed. I was just hoping that the driver would provide a bit more abstraction, so that a user wouldn't need to care how this is encoded in the hardware. So specifying the actual size would make more sense from that point of view.
But as I said, since I can't find a user, this discussion is a bit moot.

Reply:
@sdininno +1

@Amit-Radur can you please fix the prototype instead? As @sdininno said.

Reply:
@jforissier,
i had submitted the patch couple of month ago with the initial i.MX patch series.
you already had approve it .
i was planning to resubmit only this patch ( and the wdog patch) this week.

Reply:
@Andre-ARM ,
sure, user abstration can be a good thing, but this will imply a function a little bit more advanced with checks on size and alignement limitation imposed by the TZC ip. The current use of this driver is rather a static configuration at boot time so for now I just do not think it is worth the effort.